perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/GroupsManagerBlImpl.java

```diff
@@ -3030,22 +3045,104 @@ private void categorizeMembersForLightweightSynchronization(PerunSession sess, G
 		candidatesToAdd.addAll(candidates);
 		membersToRemove.addAll(groupMembers);
 		//mapping structure for more efficient searching
 		Map<UserExtSource, RichMember> mappingStructure = this.createMappingStructure(groupMembers);
 
 		//try to find already existing candidates between members in group
-		for(Candidate candidate: candidates) {
-			List<UserExtSource> candidateExtSources = candidate.getUserExtSources();
-			for(UserExtSource key: candidateExtSources) {
+		for(CandidateSync candidate: candidates) {
+			List<RichUserExtSource> candidateExtSources = candidate.getRichUserExtSources();
+			for(RichUserExtSource key: candidateExtSources) {
 				//candidate exists, will be updated
-				if(mappingStructure.containsKey(key)) {
-					membersToUpdate.put(candidate, mappingStructure.get(key));
+				RichMember richMember = processRichUserExtSourceDuringCategorization(sess, key, mappingStructure);
+				if (richMember != null) {
+					membersToUpdate.put(candidate, richMember);
 					candidatesToAdd.remove(candidate);
-					membersToRemove.remove(mappingStructure.get(key));
-					break;
+					membersToRemove.remove(richMember);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Checks if userExtSource of candidate corresponds to some userExtSource of RichMember already in group. Whether
+	 * it's by extLogin or value of ues attribute. Returns this RichMember if it does, null otherwise.
+	 *
+	 * @param sess perun session
+	 * @param key ues of candidate
+	 * @param mappingStructure map of ues to richMembers in group
+	 * @return richMember or null
+	 */
+	private RichMember processRichUserExtSourceDuringCategorization(PerunSession sess, RichUserExtSource key, Map<UserExtSource, RichMember> mappingStructure) {
+		//check if ues is in the map
+		if (mappingStructure.containsKey(key.asUserExtSource())) {
+			return mappingStructure.get(key.asUserExtSource());
+		}
+
+		for (Attribute attribute : key.getAttributes()) {
+
+			//logins are either single value string, or multi value list
+			List<String> loginsFromAttribute;
+			if (attribute.getType().equals(String.class.getName())) {
+				loginsFromAttribute = Arrays.asList(attribute.valueAsString());
+			} else {
+				loginsFromAttribute = attribute.valueAsList();
+			}
+
+			//check if extSource with login from ues attribute is in the map
+			for (UserExtSource ues : mappingStructure.keySet()) {
+				if (ues.getExtSource().equals(key.asUserExtSource().getExtSource())) {
+					try {
+						Attribute attributeToCompare = perunBl.getAttributesManagerBl().getAttribute(sess, ues, attribute.getName());
+
+						List<String> attributeToCompareValue;
+						if (attributeToCompare.getType().equals(String.class.getName())) {
+							attributeToCompareValue = Arrays.asList(attribute.valueAsString());
+						} else {
```

Comment:
At this moment, you should throw an error if there is something different than String or ArrayList (Integer, boolean, map, etc.) I am afraid this would end with some kind of unexpected Exception.